### ISSUE TITLE
CODE 6.4.10: Data Loss When Saving

### ISSUE THREAD

**Issue by gohrner** [CollaboraOnline/online#2747]
**Describe the bug**
I don't yet know how to reproduce it, but I just tried to create a testcase for Bug #2745, added lots of text to a Calc tab, duplicated the tab multiple times and closed and re-opened the Sheet - and it was empty, CODE didn't save my changes. :-(

I think I even explictly clicked the "Save" icon, but even without, the data must not be dropped / lost silently. Not every user will always click the "Save" button, especially as it wasn't necessary with earlier CODE releases, and changes were auto-saved.

Desktop (please complete the following information)

OS: Ubuntu 20.04 (KDE neon)
Browser: Chromium
Version: 91.0.4472.101 (Offizieller Build) Built on Ubuntu , running on neon 20.04 (64-Bit)

Collabora Online Development Edition (CODE)
Diese Version von Collabora Online Development Edition (CODE) wird bereitgestellt von

LOOLWSD
6.4.10 (git hash: 51b8716)
Angeboten von: ee7c0e28
LOKit
Collabora Office 6.4-42 (git hash: 8aa76ce)
"Ubuntu 18.04.5 LTS"
Copyright © 2021, Collabora Productivity Limited.

**Comment by gohrner** [comment]
I just experienced an additional case of data loss: Whly trying again to prepare a test case for #2745, I added some formulas and formatting to my sheet.

After closing / reopening it, some of the formatting I applied to some of the tables was lost.

**Comment by rpear** [comment]
I can confirm this, in my case I simply opened an existing spreadsheet, typed a few comments into a cell, clicked on another cell, closed the browsers tab without clicking Save, re-opened the the spreadsheet and the text was not there, but I could not repeat this, I tried several times. This was using 6.4.10snapshot (git hash: 545ed4a) (Commits on Jul 2, 2021) on a Chromebook web browser.

**Comment by thebearon** [comment]
The data loss is definitely concerning, unfortunately I couldn't reproduce this, either.

**Comment by gohrner** [comment]
Additional info: In contrast to previous data-loss cases I experienced and which were caused by server crashes, this time I didn't see anything crash related with CODE 6.4.10 at all.

So appears to be something else, or - in case something crashes - it's not logged.

**Comment by mmeeks** [comment]
I wonder if explicitly clicking save, and closing the tab - before the document has synched to the remote storage could be the problem (Ash?). And thanks so much for testing & filing!

**Comment by mmeeks** [comment]
One way to reproduce calc data-loss is easy:
* load the attached
* switch to tab 1
* select a cell in the middle of the screen
* set the font height to 72 type 'this is bug'
*  click the top right corner (x) to close the tab (on share) 
* come back to the document in ~5 seconds
* switch to sheet1 - it is gone - and interestingly I'm on the 4th sheet instead of the first: [testsheet.ods](https://github.com/CollaboraOnline/online/files/6771571/testsheet.ods)

**Comment by shanemarsh28** [comment]
As a new user, I can also confirm the above. On a new instance we have been unable to get .ods files to save at all. 

Additionally there is a further issue with .xls where changes are NOT saved:
1) Open .xls
2) make a change in any cell
3) open a new sheet
4) make a change in new sheet
5) "Save" the document
6) Close the document

When the document is reopened the either the new sheet is missing or the data on sheet 1 did not save. This is very alarming guys!

**Comment by mmeeks** [comment]
Hi Shane, thanks for the report - can you possibly report what is in your help->about when it comes to the hashes? We are doing a series of fixes here, some of which have been shipped and it would be great to know that precisely. (6.4.10 is not sufficient to pinpoint this). Thanks !

**Comment by shanemarsh28** [comment]
Yes of course. 

The issue is easily reproducible using the current version: Collabora Online - Built-in CODE Server for NextCloud.
LOOLWSD: 6.4.10 (git hash: 2afbc52) Served by: 4391f347
LOKit Collabora Office 6.4-44 (git hash: 2b31b2e) "Ubuntu Core 18" "Slow Proxy"

Since my last message, I replaced the built in version for this: Collabora Online Development Edition (CODE)
LOOLWSD 6.4.10 (git hash: fcb0905). Served by: 0454436c

LOKit: Collabora Office 6.4-45 (git hash: a21347f) "Ubuntu 18.04.5 LTS"

I have not been able easily reproduce the issue on #fcb0905 - it appears to be a much more stable product. We are currently monitoring.

**Comment by mmeeks** [comment]
Shane - thanks so much for that feedback; Gohrner &rRob - have you had a chance to update, and/or any more input on this? clearly its really important not to loose data during the new async-save! Thanks.

**Comment by rpear** [comment]
I can't repeat the data loss I experienced, I have also tried many variations again today on 6.4.10snapshot git hash: 25b2adb  -Commits on Jul 20, 2021. Note that I have only tested on a single ODS file.

**Comment by gohrner** [comment]
I just had data loss again:

1. Opened a document
2. Change some cells
3. Added a new sheet
4. Changed some cells
5. Added a new sheet
6. Changes some cells
7. Closed the document with the top-right-corner-(x)
8. Reopened the document => all changes gone. :-(

LOOLWSD
6.4.10 (git hash: fcb0905)
Angeboten von: 57f5f1cf
LOKit
Collabora Office 6.4-45 (git hash: a21347f)
"Ubuntu 18.04.5 LTS"

This time I had a stacktrace in the logs (I think I didn't have that last time when data was lost):

```
wsd-00010-00511 2021-07-27 14:38:23.746824 [ docbroker_014 ] WRN  Waking up dead poll thread [HttpSynReqPoll], started: false, finished: false| ./net/Socket.hpp:682
wsd-00010-00511 2021-07-27 14:38:23.788289 [ docbroker_014 ] WRN  Waking up dead poll thread [HttpSynReqPoll], started: false, finished: false| ./net/Socket.hpp:682
wsd-00010-00511 2021-07-27 14:39:02.294168 [ docbroker_014 ] WRN  Ignoring attempted read from 22| ./net/Socket.hpp:1021
kit-00469-00042 2021-07-27 14:39:02.314209 [ kitbroker_014 ] SIG   Fatal signal received: SIGSEGV
Backtrace pid: 469 - kit startup of 6.4.10 fcb0905:
/usr/bin/loolforkit(_ZN7SigUtil13dumpBacktraceEv+0x80)[0x55dd6908ffb0]
/usr/bin/loolforkit(+0x1ec7a5)[0x55dd690907a5]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12980)[0x7fce0c542980]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN3sdr7contact13ObjectContact20AddViewObjectContactERNS0_17ViewObjectContactE+0x22)[0x7fce0910a252]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN3sdr7contact17ViewObjectContactC2ERNS0_13ObjectContactERNS0_11ViewContactE+0xed)[0x7fce0911debd]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN3sdr7contact25ViewObjectContactOfSdrObjC2ERNS0_13ObjectContactERNS0_11ViewContactE+0x9)[0x7fce0911f4d9]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN3sdr7contact19ViewContactOfSdrObj37CreateObjectSpecificViewObjectContactERNS0_13ObjectContactE+0x25)[0x7fce09109575]
/opt/collaboraoffice6.4/program/../program/libsclo.so(+0xbee172)[0x7fcdf8cfb172]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZNK3sdr7contact17ViewObjectContact13getGridOffsetEv+0x50)[0x7fce0911d3a0]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZNK3sdr7contact17ViewObjectContact22getPrimitive2DSequenceERKNS0_11DisplayInfoE+0x5a8)[0x7fce0911e718]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZNK3sdr7contact17ViewObjectContact14getObjectRangeEv+0x9e)[0x7fce0911eb2e]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN3sdr7contact17ViewObjectContact13ActionChangedEv+0x22)[0x7fce0911edd2]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN3sdr7contact11ViewContact19ActionChildInsertedERS1_+0x4b)[0x7fce0910ef5b]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN10SdrObjList15NbcInsertObjectEP9SdrObjectm+0x97)[0x7fce09246467]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN10SdrObjList12InsertObjectEP9SdrObjectm+0x4f)[0x7fce0924219f]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN10FmFormPage12InsertObjectEP9SdrObjectm+0x11)[0x7fce09466c91]
/opt/collaboraoffice6.4/program/../program/libsclo.so(+0x52b9b6)[0x7fcdf86389b6]
/opt/collaboraoffice6.4/program/../program/libsclo.so(_ZNK8ScPostIt25CreateCaptionFromInitDataERK9ScAddress+0x130)[0x7fcdf863a4a0]
/opt/collaboraoffice6.4/program/../program/libsclo.so(_ZNK8ScPostIt18GetOrCreateCaptionERK9ScAddress+0x9)[0x7fcdf863b429]
/opt/collaboraoffice6.4/program/../program/libsclo.so(+0x3ff27c)[0x7fcdf850c27c]
/opt/collaboraoffice6.4/program/../program/libsclo.so(+0x549390)[0x7fcdf8656390]
/opt/collaboraoffice6.4/program/../program/libsclo.so(+0x4613ad)[0x7fcdf856e3ad]
/opt/collaboraoffice6.4/program/../program/libsclo.so(+0x805910)[0x7fcdf8912910]
/opt/collaboraoffice6.4/program/../program/libsclo.so(+0x90b004)[0x7fcdf8a18004]
/opt/collaboraoffice6.4/program/../program/libsclo.so(_ZN10ScDocShell6SaveAsER9SfxMedium+0x139)[0x7fcdf8a18429]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN14SfxObjectShell15SaveAsOwnFormatER9SfxMedium+0x108)[0x7fce08aeb5b8]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x1f1735f)[0x7fce08aee35f]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x1f18906)[0x7fce08aef906]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x1f18f5b)[0x7fce08aeff5b]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN12SfxBaseModel9storeSelfERKN3com3sun4star3uno8SequenceINS2_5beans13PropertyValueEEE+0x9f6)[0x7fce08b20d86]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x1eecb5e)[0x7fce08ac3b5e]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x1f06226)[0x7fce08add226]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN8SfxShell11ExecuteSlotER10SfxRequestPK12SfxInterface+0x4b)[0x7fce0899a96b]
/opt/collaboraoffice6.4/program/../program/libsclo.so(_ZN14ScTabViewShell11ExecuteSaveER10SfxRequest+0x74)[0x7fcdf8d7de84]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x1da8101)[0x7fce0897f101]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN13SfxDispatcher8Execute_ER8SfxShellRK7SfxSlotR10SfxRequest11SfxCallMode+0x122)[0x7fce08984582]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN13SfxDispatcher7ExecuteEt11SfxCallModePK10SfxItemSetS3_t+0x189)[0x7fce08985829]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x1df9fb2)[0x7fce089d0fb2]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN10comphelper15dispatchCommandERKN3rtl8OUStringERKN3com3sun4star3uno9ReferenceINS6_5frame6XFrameEEERKNS7_8SequenceINS6_5beans13PropertyValueEEERKNS8_INS9_23XDispatchResultListenerEEE+0x4bf)[0x7fce07fc2c2f]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN10comphelper15dispatchCommandERKN3rtl8OUStringERKN3com3sun4star3uno8SequenceINS6_5beans13PropertyValueEEERKNS7_9ReferenceINS6_5frame23XDispatchResultListenerEEE+0x168)[0x7fce07fc2f88]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x208174f)[0x7fce08c5874f]
/usr/bin/loolforkit(_ZN12ChildSession10unoCommandEPKciRK12StringVector+0x186)[0x55dd68fe83c6]
/usr/bin/loolforkit(_ZN12ChildSession12_handleInputEPKci+0x2961)[0x55dd69009cc1]
/usr/bin/loolforkit(_ZN7Session13handleMessageERKSt6vectorIcSaIcEE+0x55)[0x55dd6907ed65]
/usr/bin/loolforkit(_ZN8Document14forwardToChildERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt6vectorIcS4_E+0x746)[0x55dd69047166]
/usr/bin/loolforkit(_ZN8Document10drainQueueERKNSt6chrono10time_pointINS0_3_V212steady_clockENS0_8durationIlSt5ratioILl1ELl1000000000EEEEEE+0x260)[0x55dd6904ff10]
/usr/bin/loolforkit(_ZN13KitSocketPoll7kitPollEi+0x238)[0x55dd69052218]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN14SvpSalInstance7DoYieldEbb+0x4d3)[0x7fce09ce9a33]
/opt/collaboraoffice6.4/program/libmergedlo.so(+0x3048f8b)[0x7fce09c1ff8b]
/opt/collaboraoffice6.4/program/libmergedlo.so(_ZN11Application7ExecuteEv+0x45)[0x7fce09c20b25]
wsd-00010-00511 2021-07-27 14:39:02.366124 [ docbroker_014 ] ERR  DocumentBroker stopping although modified  Broker: commenttest.ods pid: 469 *** Marked to destroy ***
```

Shall I report this as a different issue?

**Comment by kendy** [comment]
> Shall I report this as a different issue?

@gohrner: I am afraid this is likely a different root cause, because here is clearly a crash involved - can you please report it separately?  Thank you in advance!

**Comment by kendy** [comment]
We have just released CODE 6.4.10-6 that should address this problem. Can you please re-test, and if you still see problems, either re-open it, or (better) create a new report?

Thank you in advance!
